src/mpid/ch4/netmod/ofi/ofi_am_impl.c

```diff
@@ -66,6 +67,92 @@ static int issue_deferred_am_isend_eager(MPIDI_OFI_deferred_am_isend_req_t * dre
+        /* FIXME: currently we always do packing, also for high density types. However,
+         * we should not do packing unless needed. Also, for large low-density types
+         * we should not allocate the entire buffer and do the packing at once. */
+        /* TODO: (1) Skip packing for high-density datatypes; */
+        mpi_errno = MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.pack_buf_pool,
+                                                       (void **) &send_buf);
+        if (send_buf == NULL) {
+            /* no buffer available, suppress error and exit */
+            mpi_errno = MPI_SUCCESS;
+            goto fn_exit;
+        }
+        MPIR_ERR_CHECK(mpi_errno);
+        mpi_errno = MPIR_Typerep_pack(dreq->buf, dreq->count, dreq->datatype,
+                                      MPIDIG_send_get_offset(dreq->sreq), send_buf, send_size,
+                                      &packed_size);
+        MPIR_ERR_CHECK(mpi_errno);
+        send_size = packed_size;
+
+        MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.am_hdr_buf_pool, (void **) &send_req);
+        MPIR_Assert(send_req);
+        send_req->sreq = dreq->sreq;
+        if (dreq->am_hdr_sz > 0) {
+            send_req->am_hdr_sz = dreq->am_hdr_sz;
+            send_req->am_hdr = (void *) &send_req->am_hdr_buf[0];
+            MPIR_Memcpy(send_req->am_hdr, dreq->am_hdr, dreq->am_hdr_sz);
+        } else {
+            send_req->am_hdr_sz = 0;
+        }
+        send_req->pack_buffer = send_buf;
+    } else {
+        /* TODO: we will need to handle the unpacked dense noncontig data here */
+        MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.am_hdr_buf_pool, (void **) &send_req);
+        MPIR_Assert(send_req);
+        send_req->sreq = dreq->sreq;
+        if (dreq->am_hdr_sz > 0) {
+            send_req->am_hdr_sz = dreq->am_hdr_sz;
+            send_req->am_hdr = (void *) &send_req->am_hdr_buf[0];
+            MPIR_Memcpy(send_req->am_hdr, dreq->am_hdr, dreq->am_hdr_sz);
+        } else {
+            send_req->am_hdr_sz = 0;
+        }
+        send_req->pack_buffer = NULL;
+        MPIDI_Datatype_get_ptr_lb(dreq->count, dreq->datatype, dt_ptr, dt_true_lb);
+        send_buf = (char *) (dreq->buf) + dt_true_lb + MPIDIG_send_get_offset(dreq->sreq);
+    }
+
+    MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                    (MPL_DBG_FDEST,
+                     "issue pipeline seg for req handle=0x%x send_size %ld", dreq->sreq->handle,
+                     send_size));
+
+    mpi_errno = MPIDI_OFI_am_isend_pipeline(dreq->rank, dreq->comm, dreq->handler_id, send_buf,
+                                            send_size, dreq->data_sz, dreq->sreq, send_req);
+    MPIDIG_send_issue_seg(dreq->sreq, send_size);
+    /* if we reach here, the first segment must have been sent. set am_hdr_sz to 0 to skip am_hdr
+     * on the rest segments */
+    dreq->am_hdr_sz = 0;
+    if (MPIDIG_send_get_data_sz_left(dreq->sreq) == 0) {
```

Comment:
See note earlier on changing it to `MPDIG_send_is_done` (or something like that).

Reply:
OK